### PR TITLE
Make OptimismConfig.EIP1559DenominatorCanyon optional

### PR DESCRIPTION
**Pull request by bitwiseguy**
Prior to this PR the `OptimismConfig.EIP1559DenominatorCanyon` field was required, so when Unmarshaled from a json file in the superchain-registry, it was populated with the default `0` value if it was missing. 

This PR makes that field optional, so that it is initialized as `nil`, and only populated if it is present in the json file. This allows the new superchain-registry `add-chain check-genesis` command to be backwards compatible with chains that did not have canyon activated at genesis.

### Additional context
* superchain-registry PR to add the `add-chain check-genesis` command: https://github.com/ethereum-optimism/superchain-registry/pull/354

### CODE REVIEW

File: core/genesis.go [hunk 1]
```diff
@@ -282,8 +282,9 @@ func SetupGenesisBlockWithOverride(db ethdb.Database, triedb *triedb.Database, g
 			if overrides != nil && overrides.OverrideOptimismCanyon != nil {
 				config.CanyonTime = overrides.OverrideOptimismCanyon
 				config.ShanghaiTime = overrides.OverrideOptimismCanyon
-				if config.Optimism != nil && config.Optimism.EIP1559DenominatorCanyon == 0 {
-					config.Optimism.EIP1559DenominatorCanyon = 250
+				if config.Optimism != nil && *config.Optimism.EIP1559DenominatorCanyon == 0 {
+					eip1559DenominatorCanyon := uint64(250)
+					config.Optimism.EIP1559DenominatorCanyon = &eip1559DenominatorCanyon
```
Comment (geoknee) [inline core/genesis.go]: Shouldn't we check that `*config.Optimism.EIP1559DenominatorCanyon != nil` as well, or risk a panic here?

File: params/config.go [hunk 2]
```diff
@@ -898,7 +898,7 @@ func (c *ChainConfig) checkCompatible(newcfg *ChainConfig, headNumber *big.Int,
 func (c *ChainConfig) BaseFeeChangeDenominator(time uint64) uint64 {
 	if c.Optimism != nil {
 		if c.IsCanyon(time) {
-			return c.Optimism.EIP1559DenominatorCanyon
+			return *c.Optimism.EIP1559DenominatorCanyon
```
Comment (bitwiseguy) [inline params/config.go:904]: What is the best way to avoid a potential runtime panic here? Is `EIP1559DenominatorCanyon` guaranteed to be non-nil if it hits this part of the code, i.e. `if c.IsCanyon(time)`?

File: core/genesis.go (hunk 1 above)
Comment (bitwiseguy) [reply core/genesis.go]: Changed the code so that if the `overrides.OverrideOptimismCanyon` field is set, this value will always get set to 250

File: core/genesis.go [hunk 3]
```diff
@@ -282,8 +282,9 @@ func SetupGenesisBlockWithOverride(db ethdb.Database, triedb *triedb.Database, g
 			if overrides != nil && overrides.OverrideOptimismCanyon != nil {
 				config.CanyonTime = overrides.OverrideOptimismCanyon
 				config.ShanghaiTime = overrides.OverrideOptimismCanyon
-				if config.Optimism != nil && config.Optimism.EIP1559DenominatorCanyon == 0 {
-					config.Optimism.EIP1559DenominatorCanyon = 250
+				if config.Optimism != nil {
+					eip1559DenominatorCanyon := uint64(250)
+					config.Optimism.EIP1559DenominatorCanyon = &eip1559DenominatorCanyon
```
Comment (sebastianst) [inline core/genesis.go]: Isn't this changing the old logic? The old logic said, only set the default value of 250 if the current value is `0`, which indicates an unset value. But it might already be non-zero. So instead, shouldn't this now be:
```go
if config.Optimism != nil && (config.Optimism.EIP1559DenominatorCanyon == nil || *config.Optimism.EIP1559DenominatorCanyon == 0) {
```
?

Also wondering if there's an existing test where we could add the three cases `nil`, `0`, and _non-zero_ to?

File: params/superchain.go [hunk 4]
```diff
@@ -76,7 +77,7 @@ func LoadOPStackChainConfig(chainID uint64) (*ChainConfig, error) {
 		Optimism: &OptimismConfig{
 			EIP1559Elasticity:        6,
 			EIP1559Denominator:       50,
-			EIP1559DenominatorCanyon: 250,
+			EIP1559DenominatorCanyon: &eip1559DenominatorCanyon,
```
Comment (sebastianst) [inline params/superchain.go]: you could also use
```suggestion
			EIP1559DenominatorCanyon: newUint64(250),
```

File: params/config.go (hunk 2 above)
Comment (sebastianst) [reply params/config.go:904]: Good question. A properly set up chain config that has Canyon activation must have a value here. This is also guaranteed by `applyOverrides` within `SetupGenesisBlockWithOverride`, which is called by a geth node starting up. So I think we're good. We also didn't check for it not being zero before, so we might just leave it as is.
However, if there are any code paths, probably only possible from within tests, that reach this point with a nil or zero Canyon denominator, it's probably worth panicking. So wdyt about adding an explicit check here for nil or zero and then panic with a message about this?

Comment (bitwiseguy) [reply params/config.go:904]: Yea I think it makes sense to add a panic to make it obvious when an invalid value is used

File: core/genesis.go (hunk 3 above)
Comment (bitwiseguy) [reply core/genesis.go]: Yes that was changing the logic. I understand now the override should only take effect when the value is unset so will accept your suggestion. Will look for tests to modify and check for the three cases

Comment (bitwiseguy) [reply core/genesis.go]: Tests added here https://github.com/ethereum-optimism/op-geth/pull/345/commits/682f7bbefe61efbc813e01cd7b64821cb61ac128

File: core/genesis.go (hunk 1 above)
Comment (bitwiseguy) [reply core/genesis.go]: Changed logic here https://github.com/ethereum-optimism/op-geth/pull/345/commits/a850e59f45a9f1ad81dd7faf72214dc89f840088

File: params/config.go (hunk 2 above)
Comment (bitwiseguy) [reply params/config.go:904]: panic added here https://github.com/ethereum-optimism/op-geth/pull/345/commits/c3bcfeb1e56b001c097d432e5d66a47b6726039b

File: core/superchain_test.go [hunk 5]
```diff
@@ -37,17 +37,56 @@ func TestRegistryChainConfigOverride(t *testing.T) {
...
 	}
-	// check if we have a corrected chain config
-	if chainConfig.RegolithTime == nil {
-		t.Fatal("expected regolith time to be corrected, but time is still nil")
-	} else if *chainConfig.RegolithTime != expectedRegolithTime {
-		t.Fatalf("expected regolith time to be %d, but got %d", expectedRegolithTime, *chainConfig.RegolithTime)
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			genesis.Config.Optimism.EIP1559DenominatorCanyon = tt.setDenominator
```
Comment (bitwiseguy) [inline core/superchain_test.go:80]: Do we need to make a deep copy of `genesis` here instead of modifying the variable share across all subtests?

Comment (geoknee) [reply core/superchain_test.go:80]: We should prepare a fresh `genesis` in each sub test, so we could move the setup code above inside the `t.Run` callback.

Comment (bitwiseguy) [reply core/superchain_test.go:80]: https://github.com/ethereum-optimism/op-geth/pull/345/commits/8aca18b466a72918a7f0be540e0c633a6fadd44f

**Review by sebastianst — APPROVED** [review]
